src/empirical.jl

```diff
@@ -51,18 +51,43 @@ struct EmpiricalVariogram
     sums   = zeros(nlags)
     counts = zeros(Int, nlags)
 
+    # decide whether to construct a spatial index
+    # coordinates need to be AbstractFloat for NearestNeighbors to work
+    # https://github.com/KristofferC/NearestNeighbors.jl/issues/13
+    isfloat = eltype(coordinates(sdata)) <: AbstractFloat
+    # quite arbitrary, but for small datasets setting up the index is too much overhead
+    # it really also depends on what fraction of points are within hmax
+    isbig = npts > 1000
+
+    if isfloat && isbig
+      neigh = BallNeighborhood(hmax, distance)
+      neighsearcher = NeighborhoodSearcher(sdata, neigh)
+      use_spatial_index = true
+    else
+      use_spatial_index = false
+    end
+
     # preallocate memory for coordinates
     xi = MVector{ndims(sdata),coordtype(sdata)}(undef)
     xj = MVector{ndims(sdata),coordtype(sdata)}(undef)
 
-    @inbounds (
-    for j in 1:npts
+    @inbounds for j in 1:npts
       coordinates!(xj, sdata, j)
-      for i in j+1:npts
+
+      if use_spatial_index
+        pt_indices = search(xj, neighsearcher)
+      else
+        pt_indices = j+1:npts
+      end
+
+      for i in pt_indices
+        i <= j && continue  # spatial index include these
```

Comment:
This check for example should only exist in the branch that computes the variogram with neighbourhood search.

Reply:
You're right, I did this in 0009ea940c63a829c0238850f89467d7e5bf8557. Don't think it's an expensive check, but indeed not needed.